pkgs/native_assets_cli/lib/src/code_assets/validation.dart

```diff
@@ -5,55 +5,60 @@
 import 'dart:io';
 
 import '../../code_assets_builder.dart';
-
+import '../json_utils.dart';
+import 'config.dart';
 import 'link_mode.dart';
 
+// TODO: The validation should work on the JSON, not on the parsed config.
 Future<ValidationErrors> validateCodeAssetBuildConfig(
         BuildConfig config) async =>
     _validateCodeConfig(
       'BuildConfig',
-      config.codeConfig.targetOS,
       // ignore: deprecated_member_use_from_same_package
       config.dryRun,
-      config.codeConfig,
+      config.json,
     );
 
 Future<ValidationErrors> validateCodeAssetLinkConfig(LinkConfig config) async =>
-    _validateCodeConfig(
-        'LinkConfig', config.codeConfig.targetOS, false, config.codeConfig);
+    _validateCodeConfig('LinkConfig', false, config.json);
 
 ValidationErrors _validateCodeConfig(
-    String configName, OS targetOS, bool dryRun, CodeConfig codeConfig) {
+    String configName, bool dryRun, Map<String, Object?> json) {
   // The dry run will be removed soon.
   if (dryRun) return const [];
 
   final errors = <String>[];
+  final targetOSJson = json[targetOSConfigKey] as String?;
+  final targetOS = targetOSJson == null ? null : OS.fromString(targetOSJson);
```

Comment:
I'm not particularly fond of these changes here. These changes occur because
* the underlying json schema makes the macos/android/ios/... configs optional, but they have to be present depending on OS
* we chose an API that makes them non-nullable

The second point: This pattern with non-nullable sub-fields (which are nullable on the protocol level) that should only be used under guard conditions usually have `has*` getters. If we followed that style here as well, we'd have user code like this:

```dart
final codeConfig = buildConfig.codeConfig;
if (codeConfig.hasMacOSConfig) { // <-- could also use `codeConfig.targetOS == OS.macos`
  use(codeConfig.macOSConfig);  // <-- use non-nullable API
}
```

The validation would then look like this
```dart
switch (codeConfig.targetOS) {
  case OS.macOS:
    if (!codeConfig.hasMacOSConfig)) {
      errors.add('...');
    }
    break;
  ...
}
```

That would make us not have to deal with JSON here at all. It would all be encapsulated in the `CodeConfig`/... objects.